fix non-persistent buffers

avikchaudhuri opened the pull request:
Summary:
Dynamo doesn't track whether buffers are `persistent`. This led to some ugly code where we would mark buffers as always persistent when creating signatures, then later check whether the buffers were not in the state dict to infer whether they were non-persistent, and use this to fix up the signature.

This PR instead defines a utility to look up all the non-persistent buffers registered inside a module (this information is recorded in a private `_non_persistent_buffers_set` module attribute), and uses it to (a) correctly set the persistent flag on buffers when creating signatures (b) transfer this information to a Dynamo-traced graph module, which then causes non-persistent buffers to (correctly) not show up in the state dict.

Test Plan: existing tests + new case with non-persistent buffer in nested module

Differential Revision: D60224656

zhxchen17 reviewed (approved):
lgtm.

one lint error

ydwu4 commented on torch/export/_trace.py:
```diff
@@ -1279,11 +1287,21 @@ def _strict_export_lower_to_aten_ir(
     constant_attrs = _gather_constant_attrs(mod)
+    param_buffer_table: Dict[str, str] = _get_param_buffer_mapping(mod, gm_torch_level)
+
+    # Dynamo does not track which buffers were registered as non-persistent. This info
```
How about turn L1291-1300 into a standalone function something like _set_non_persistent_buffers or something?

avikchaudhuri replied:
I considered doing that, but `param_buffer_table` is also used later for restoring names.

So if we have to name this function, it feels like another kind of `restore` function. But above we explicitly decided not to restore names in the graph module for some reason (see `restore_fqn` argument) so the whole workflow feels like a hodge-podge. I think the good thing to do long-term would be to understand why we don't do all the restoration in one place.

ydwu4 reviewed (approved):
Looks great!

avikchaudhuri committed:
fix non-persistent buffers (#131756)

Summary:
Pull Request resolved: https://github.com/pytorch/pytorch/pull/131756

Dynamo doesn't track whether buffers are `persistent`. This led to some ugly code where we would mark buffers as always persistent when creating signatures, then later check whether the buffers were not in the state dict to infer whether they were non-persistent, and use this to fix up the signature.

This PR instead defines a utility to look up all the non-persistent buffers registered inside a module (this information is recorded in a private `_non_persistent_buffers_set` module attribute), and uses it to (a) correctly set the persistent flag on buffers when creating signatures (b) transfer this information to a Dynamo-traced graph module, which then causes non-persistent buffers to (correctly) not show up in the state dict.

Test Plan: existing tests + new case with non-persistent buffer in nested module

Reviewed By: ydwu4, zhxchen17

Differential Revision: D60224656
